In [1]:
import pandas as pd
import os

C:\Users\kusha\AppData\Local\Temp\ipykernel_9996\3026406390.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [19]:
# loading datasets to perform EDA

df_economical = pd.read_csv('csv_datasets\most_economical_2008_2023.csv')
df_wicket = pd.read_csv('csv_datasets\most_wicket_2008_2023.csv')

# Most Economical Bowlers (2008-2023)

In [20]:
df_economical.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  61 non-null     int64  
 1   Player      61 non-null     object 
 2   Team(s)     60 non-null     object 
 3   Mat         60 non-null     float64
 4   Overs       60 non-null     float64
 5   Mdns        60 non-null     float64
 6   Runs        60 non-null     float64
 7   Wkts        60 non-null     float64
 8   Avg         60 non-null     float64
 9   Best        60 non-null     object 
 10  E/R         60 non-null     float64
 11  S/R         60 non-null     float64
 12  4w          60 non-null     float64
dtypes: float64(9), int64(1), object(3)
memory usage: 6.3+ KB


In [21]:
df_economical.head(2)

Unnamed: 0       Player Team(s)   Mat  Overs  Mdns    Runs  Wkts    Avg  \
0           0     A Kumble     RCB  42.0  160.5   1.0  1058.0  45.0  23.51   
1           1  G D McGrath     DDV  14.0   54.0   2.0   357.0  12.0  29.75   

   Best   E/R    S/R   4w  
0   5/5  6.58  21.44  3.0  
1  4/29  6.61  27.00  1.0

- Remove redundant columns and dropping record with null values.

In [22]:
df_economical.drop(columns=['Unnamed: 0'], inplace = True)
df_economical.dropna(axis=0, inplace=True)

In [24]:
df_economical.info()

<class 'pandas.core.frame.DataFrame'>
Index: 60 entries, 0 to 59
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Player   60 non-null     object 
 1   Team(s)  60 non-null     object 
 2   Mat      60 non-null     float64
 3   Overs    60 non-null     float64
 4   Mdns     60 non-null     float64
 5   Runs     60 non-null     float64
 6   Wkts     60 non-null     float64
 7   Avg      60 non-null     float64
 8   Best     60 non-null     object 
 9   E/R      60 non-null     float64
 10  S/R      60 non-null     float64
 11  4w       60 non-null     float64
dtypes: float64(9), object(3)
memory usage: 6.1+ KB


# Bowlers with most wicket (2008-2023)

In [25]:
df_wicket.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  61 non-null     int64  
 1   Player      61 non-null     object 
 2   Team(s)     60 non-null     object 
 3   Mat         60 non-null     float64
 4   Overs       60 non-null     float64
 5   Mdns        60 non-null     float64
 6   Runs        60 non-null     float64
 7   Wkts        60 non-null     float64
 8   Avg         60 non-null     float64
 9   Best        60 non-null     object 
 10  E/R         60 non-null     float64
 11  S/R         60 non-null     float64
 12  4w          60 non-null     float64
dtypes: float64(9), int64(1), object(3)
memory usage: 6.3+ KB


- Remove redundant columns and dropping record with null values.

In [26]:
df_wicket.head(2)

Unnamed: 0       Player        Team(s)    Mat  Overs  Mdns    Runs   Wkts  \
0           0   Y S Chahal  MIN, RCB, RRO  145.0  528.5   4.0  4056.0  187.0   
1           1  D J J Bravo  MIN, CSK, GUJ  161.0  519.5   3.0  4360.0  183.0   

     Avg  Best   E/R    S/R   4w  
0  21.69  5/40  7.67  16.97  7.0  
1  23.83  4/22  8.39  17.04  2.0

In [27]:
df_wicket.drop(columns=['Unnamed: 0'], inplace = True)
df_wicket.dropna(axis=0, inplace=True)

# Merging datasets

In [28]:
df_wicket.info()

<class 'pandas.core.frame.DataFrame'>
Index: 60 entries, 0 to 59
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Player   60 non-null     object 
 1   Team(s)  60 non-null     object 
 2   Mat      60 non-null     float64
 3   Overs    60 non-null     float64
 4   Mdns     60 non-null     float64
 5   Runs     60 non-null     float64
 6   Wkts     60 non-null     float64
 7   Avg      60 non-null     float64
 8   Best     60 non-null     object 
 9   E/R      60 non-null     float64
 10  S/R      60 non-null     float64
 11  4w       60 non-null     float64
dtypes: float64(9), object(3)
memory usage: 6.1+ KB


In [29]:
df_economical.info()

<class 'pandas.core.frame.DataFrame'>
Index: 60 entries, 0 to 59
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Player   60 non-null     object 
 1   Team(s)  60 non-null     object 
 2   Mat      60 non-null     float64
 3   Overs    60 non-null     float64
 4   Mdns     60 non-null     float64
 5   Runs     60 non-null     float64
 6   Wkts     60 non-null     float64
 7   Avg      60 non-null     float64
 8   Best     60 non-null     object 
 9   E/R      60 non-null     float64
 10  S/R      60 non-null     float64
 11  4w       60 non-null     float64
dtypes: float64(9), object(3)
memory usage: 6.1+ KB


In [32]:
# Merge the dataframes based on the index
df_bowlers = pd.merge(df_wicket, df_economical, how='outer')
df_bowlers

Player                       Team(s)    Mat  Overs  Mdns  \
0           A B Dinda       KKR, DDV, PWI, RCB, RPS   78.0  252.4   6.0   
1     A D Mascarenhas                      RRO, KXI   13.0   51.2   1.0   
2         A D Russell                      DDV, KKR  112.0  253.3   0.0   
3            A Kumble                           RCB   42.0  160.5   1.0   
4            A Mishra            DDV, DCH, SUN, LSG  161.0  559.5   6.0   
..                ...                           ...    ...    ...   ...   
90  Washington Sundar                 RPS, RCB, SUN   58.0  170.4   1.0   
91         Y K Pathan                 RRO, KKR, SUN  174.0  189.1   1.0   
92         Y S Chahal                 MIN, RCB, RRO  145.0  528.5   4.0   
93       Yuvraj Singh  KXI, PWI, RCB, DDV, SUN, MIN  132.0  144.5   0.0   
94             Z Khan                 RCB, MIN, DDV  100.0  366.4   5.0   

      Runs   Wkts    Avg  Best   E/R    S/R   4w  
0   2073.0   69.0  30.04  4/18  8.20  21.97  1.0  
1    356.0   19.0  18.74  5/25  6.94  16.21  1.0  
2   2351.0   96.0  24.49  5/15  9.27  15.84  3.0  
3   1058.0   45.0  23.51   5/5  6.58  21.44  3.0  
4   4129.0  173.0  23.87  5/17  7.38  19.42  5.0  
..     ...    ...    ...   ...   ...    ...  ...  
90  1252.0   36.0  34.78  3/16  7.34  28.44  0.0  
91  1387.0   42.0  33.02  3/20  7.33  27.02  0.0  
92  4056.0  187.0  21.69  5/40  7.67  16.97  7.0  
93  1077.0   36.0  29.92  4/29  7.44  24.14  2.0  
94  2782.0  102.0  27.27  4/17  7.59  21.57  1.0  

[95 rows x 12 columns]

In [33]:
df_bowlers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95 entries, 0 to 94
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Player   95 non-null     object 
 1   Team(s)  95 non-null     object 
 2   Mat      95 non-null     float64
 3   Overs    95 non-null     float64
 4   Mdns     95 non-null     float64
 5   Runs     95 non-null     float64
 6   Wkts     95 non-null     float64
 7   Avg      95 non-null     float64
 8   Best     95 non-null     object 
 9   E/R      95 non-null     float64
 10  S/R      95 non-null     float64
 11  4w       95 non-null     float64
dtypes: float64(9), object(3)
memory usage: 9.0+ KB


## KPIs / Inferences

Q. Most economical bowlers.
<br>Q. Top wicket taking bowlers.
<br>Q. Most valuable bowlers with economy and wickets above average among the others top bowlers.
<br>Q. Bowlers for death overs among the top-performers. ( Ability to take wickets as well bowl maiden overs).

# Economy
A bowler's economy rate, often referred to simply as "economy," is a key metric in cricket that measures the average number of runs conceded by the bowler per over bowled. It is a crucial indicator of a bowler's efficiency and effectiveness in restricting the scoring rate of the opposing batsmen.

Here's how the economy rate is calculated:

Economy Rate = (Total Runs Conceded / Total Overs Bowled)

# Most Economical Bowlers

In [34]:
df_bowlers.describe()

Mat       Overs       Mdns         Runs        Wkts        Avg  \
count   95.000000   95.000000  95.000000    95.000000   95.000000  95.000000   
mean    82.526316  259.928421   2.684211  2006.747368   74.031579  28.480632   
std     48.858642  153.070082   2.821887  1168.340940   46.462168   6.075180   
min     13.000000   51.200000   0.000000   356.000000    8.000000  18.730000   
25%     43.500000  145.000000   1.000000  1081.500000   36.000000  24.545000   
50%     75.000000  247.200000   2.000000  1862.000000   69.000000  27.540000   
75%    107.000000  336.200000   3.000000  2677.000000   96.500000  30.340000   
max    226.000000  699.000000  14.000000  4902.000000  187.000000  55.630000   

             E/R        S/R         4w  
count  95.000000  95.000000  95.000000  
mean    7.659684  22.461684   1.431579  
std     0.599561   5.342907   1.692616  
min     6.580000  14.770000   0.000000  
25%     7.265000  19.155000   0.000000  
50%     7.560000  21.340000   1.000000  
75%     8.085000  24.360000   2.000000  
max     9.270000  48.750000   8.000000

- Wickets (Wkts):

On average, bowlers have taken approximately 74 wickets.
There is significant variability in wicket-taking abilities among bowlers, as indicated by the standard deviation of around 46.46.
The range of wickets taken spans from a minimum of 8 to a maximum of 187, highlighting the varying effectiveness of bowlers in taking wickets.

- Bowling Average (Avg):

The average bowling average among bowlers is about 28.48.
Bowling averages range from a minimum of 18.73 to a maximum of 55.63.
A lower bowling average generally indicates better performance, suggesting that, on average, the bowlers have performed reasonably well in terms of average.

- Economy Rate (E/R):

The average economy rate among bowlers is approximately 7.66.
There is limited variability in economy rates among bowlers, with a standard deviation of 0.60.
A lower economy rate is typically desirable, indicating that the bowlers have been relatively economical in conceding runs.

- Strike Rate (S/R):

The average strike rate among bowlers is about 22.46.
Strike rates range from a minimum of 14.77 to a maximum of 48.75.
A lower strike rate suggests that bowlers take wickets more frequently, which is generally indicative of better bowling performance.

In [45]:
average_eco = df_bowlers['E/R'].mean()
most_economical = df_bowlers[df_bowlers['E/R']<=average_eco]
most_economical[['Player','E/R']]

Player   E/R
1     A D Mascarenhas  6.94
3            A Kumble  6.58
4            A Mishra  7.38
6           A R Patel  7.24
8        Ankit Sharma  7.26
9             B Kumar  7.39
10              B Lee  7.51
13          C V Varun  7.45
14      D E Bollinger  7.22
17        D L Vettori  6.78
18         D P Nannes  7.43
20          D W Steyn  6.92
21           G B Hogg  7.47
22        G D McGrath  6.61
24    Harbhajan Singh  7.11
29      Iqbal Abdulla  7.23
31            J Botha  6.92
32         J C Archer  7.43
35         J J Bumrah  7.40
37            J Yadav  6.85
38         J-P Duminy  7.38
40         K H Pandya  7.33
46          M A Starc  7.17
47       M F Maharoof  7.43
50        M J Santner  6.88
51           M Kartik  7.24
52            M M Ali  6.95
53          M M Patel  7.52
56     M Muralitharan  6.63
57      Mohammad Nabi  7.14
63           P P Ojha  7.37
64         R A Jadeja  7.60
65           R Ashwin  7.01
66           R Bhatia  7.33
67         R D Chahar  7.56
68  R E van der Merwe  6.74
69         R J Harris  7.55
71          R R Powar  7.42
72           R Sharma  7.02
74        Rashid Khan  6.67
75       Ravi Bishnoi  7.59
76       S K  Trivedi  7.59
77          S K Raina  7.42
78          S K Warne  7.27
80        S L Malinga  7.16
82           S Nadeem  7.56
83         S P Narine  6.73
87    Shakib Al Hasan  7.44
90  Washington Sundar  7.34
91         Y K Pathan  7.33
93       Yuvraj Singh  7.44
94             Z Khan  7.59

In [46]:
len(most_economical)

52

- Total 52 economical bowlers across all seasons.

# Most Wicket Taking Bowlers

In [43]:
average_wicket = df_bowlers['Wkts'].mean()
most_wicket = df_bowlers[df_bowlers['Wkts']>=average_wicket]
most_wicket[['Player','Wkts']]

Player   Wkts
2       A D Russell   96.0
4          A Mishra  173.0
5           A Nehra  106.0
6         A R Patel  112.0
9           B Kumar  170.0
12       C H Morris   95.0
15      D J J Bravo  183.0
19     D S Kulkarni   86.0
20        D W Steyn   97.0
23        H V Patel  111.0
24  Harbhajan Singh  150.0
26       I K Pathan   80.0
27         I Sharma   82.0
28      Imran Tahir   82.0
30       J A Morkel   85.0
33      J D Unadkat   91.0
35       J J Bumrah  145.0
42         K Rabada  106.0
45         L Balaji   76.0
54       M M Sharma  119.0
55         M Morkel   77.0
58   Mohammed Shami  127.0
59   Mohammed Siraj   78.0
61          P Kumar   90.0
62       P P Chawla  179.0
63         P P Ojha   89.0
64       R A Jadeja  152.0
65         R Ashwin  171.0
70        R P Singh   90.0
73    R Vinay Kumar  105.0
74      Rashid Khan  139.0
80      S L Malinga  170.0
81       S N Thakur   89.0
83       S P Narine  163.0
84       S R Watson   92.0
86   Sandeep Sharma  124.0
88        T A Boult  105.0
89        U T Yadav  136.0
92       Y S Chahal  187.0
94           Z Khan  102.0

In [44]:
len(most_wicket)

40

- 40 bowlers with high wicket taking capacity.

# Most Valuable bowler with more than average wickets and economy

In [48]:
most_valuable = df_bowlers[(df_bowlers['Wkts']>average_wicket) & (df_bowlers['E/R']<average_eco)]
most_valuable

Player                  Team(s)    Mat  Overs  Mdns    Runs  \
4          A Mishra       DDV, DCH, SUN, LSG  161.0  559.5   6.0  4129.0   
6         A R Patel                 KXI, DDV  136.0  472.3   2.0  3421.0   
9           B Kumar                 PWI, SUN  160.0  594.4  11.0  4396.0   
20        D W Steyn       RCB, DCH, SUN, GUJ   95.0  362.4   7.0  2508.0   
24  Harbhajan Singh            MIN, CSK, KKR  163.0  569.2   6.0  4046.0   
35       J J Bumrah                      MIN  120.0  457.0   7.0  3380.0   
63         P P Ojha                 DCH, MIN   92.0  316.3   1.0  2332.0   
64       R A Jadeja       RRO, KTK, CSK, GUJ  226.0  591.1   2.0  4495.0   
65         R Ashwin  CSK, RPS, KXI, DDV, RRO  197.0  699.0   6.0  4902.0   
74      Rashid Khan                 SUN, GTI  109.0  432.5   4.0  2885.0   
80      S L Malinga                      MIN  122.0  471.1   8.0  3374.0   
83       S P Narine                      KKR  162.0  624.1   3.0  4203.0   
94           Z Khan            RCB, MIN, DDV  100.0  366.4   5.0  2782.0   

     Wkts    Avg  Best   E/R    S/R   4w  
4   173.0  23.87  5/17  7.38  19.42  5.0  
6   112.0  30.54  4/21  7.24  25.31  1.0  
9   170.0  25.86  5/19  7.39  20.99  4.0  
20   97.0  25.86   3/8  6.92  22.43  0.0  
24  150.0  26.97  5/18  7.11  22.77  2.0  
35  145.0  23.31  5/10  7.40  18.91  3.0  
63   89.0  26.20  3/11  7.37  21.34  0.0  
64  152.0  29.57  5/16  7.60  23.34  4.0  
65  171.0  28.67  4/34  7.01  24.53  1.0  
74  139.0  20.76  4/24  6.67  18.68  2.0  
80  170.0  19.85  5/13  7.16  16.63  7.0  
83  163.0  25.79  5/19  6.73  22.98  8.0  
94  102.0  27.27  4/17  7.59  21.57  1.0

In [49]:
len(most_valuable)

13

- 13 most valuable bowlers with lower economy and high wicket taking ability.

# Bowlers for Death-Overs

In [50]:
avg_maiden = df_bowlers['Mdns'].mean()
death_over_players = df_bowlers[ (df_bowlers['Mdns']>=avg_maiden) & (df_bowlers['Wkts']>=average_wicket)]
death_over_players

Player                            Team(s)    Mat  Overs  Mdns  \
4          A Mishra                 DDV, DCH, SUN, LSG  161.0  559.5   6.0   
9           B Kumar                           PWI, SUN  160.0  594.4  11.0   
15      D J J Bravo                      MIN, CSK, GUJ  161.0  519.5   3.0   
19     D S Kulkarni                      MIN, RRO, GUJ   92.0  297.5   8.0   
20        D W Steyn                 RCB, DCH, SUN, GUJ   95.0  362.4   7.0   
24  Harbhajan Singh                      MIN, CSK, KKR  163.0  569.2   6.0   
26       I K Pathan            KXI, DDV, SUN, RPS, GUJ  103.0  340.3  11.0   
27         I Sharma       KKR, DCH, SUN, RPS, KXI, DDV  101.0  357.5   6.0   
30       J A Morkel                 CSK, RCB, DDV, RPS   91.0  287.1   3.0   
33      J D Unadkat  KKR, RCB, DDV, RPS, RRO, MIN, LSG   94.0  325.0   3.0   
35       J J Bumrah                                MIN  120.0  457.0   7.0   
55         M Morkel                      RRO, DDV, KKR   70.0  271.3   3.0   
58   Mohammed Shami                 KKR, DDV, KXI, GTI  110.0  404.2   3.0   
59   Mohammed Siraj                           SUN, RCB   79.0  272.2   4.0   
61          P Kumar            RCB, KXI, MIN, SUN, GUJ  119.0  420.4  14.0   
65         R Ashwin            CSK, RPS, KXI, DDV, RRO  197.0  699.0   6.0   
74      Rashid Khan                           SUN, GTI  109.0  432.5   4.0   
80      S L Malinga                                MIN  122.0  471.1   8.0   
81       S N Thakur            KXI, RPS, CSK, DDV, KKR   86.0  279.2   3.0   
83       S P Narine                                KKR  162.0  624.1   3.0   
84       S R Watson                      RRO, RCB, CSK  145.0  338.1   3.0   
86   Sandeep Sharma                      KXI, SUN, RRO  116.0  432.2   8.0   
88        T A Boult            SUN, KKR, DDV, MIN, RRO   88.0  336.1  11.0   
89        U T Yadav                      DDV, KKR, RCB  141.0  484.2   5.0   
92       Y S Chahal                      MIN, RCB, RRO  145.0  528.5   4.0   
94           Z Khan                      RCB, MIN, DDV  100.0  366.4   5.0   

      Runs   Wkts    Avg  Best   E/R    S/R   4w  
4   4129.0  173.0  23.87  5/17  7.38  19.42  5.0  
9   4396.0  170.0  25.86  5/19  7.39  20.99  4.0  
15  4360.0  183.0  23.83  4/22  8.39  17.04  2.0  
19  2474.0   86.0  28.77  4/14  8.31  20.78  1.0  
20  2508.0   97.0  25.86   3/8  6.92  22.43  0.0  
24  4046.0  150.0  26.97  5/18  7.11  22.77  2.0  
26  2649.0   80.0  33.11  3/24  7.78  25.54  0.0  
27  2911.0   82.0  35.50  5/12  8.14  26.18  1.0  
30  2359.0   85.0  27.75  4/32  8.21  20.27  1.0  
33  2872.0   91.0  31.56  5/25  8.84  21.43  2.0  
35  3380.0  145.0  23.31  5/10  7.40  18.91  3.0  
55  2089.0   77.0  27.13  4/20  7.69  21.16  1.0  
58  3412.0  127.0  26.87  4/11  8.44  19.10  2.0  
59  2326.0   78.0  29.82  4/21  8.54  20.95  2.0  
61  3251.0   90.0  36.12  3/18  7.73  28.04  0.0  
65  4902.0  171.0  28.67  4/34  7.01  24.53  1.0  
74  2885.0  139.0  20.76  4/24  6.67  18.68  2.0  
80  3374.0  170.0  19.85  5/13  7.16  16.63  7.0  
81  2560.0   89.0  28.76  4/36  9.16  18.83  1.0  
83  4203.0  163.0  25.79  5/19  6.73  22.98  8.0  
84  2682.0   92.0  29.15  4/29  7.93  22.05  1.0  
86  3398.0  124.0  27.40  4/20  7.86  20.92  2.0  
88  2787.0  105.0  26.54  4/18  8.29  19.21  1.0  
89  4086.0  136.0  30.04  4/23  8.44  21.37  3.0  
92  4056.0  187.0  21.69  5/40  7.67  16.97  7.0  
94  2782.0  102.0  27.27  4/17  7.59  21.57  1.0

In [51]:
len(death_over_players)

26

- Total 26 bowlers with potential to perform well in death-overs.